---
title: "EWAS minfi QC 3 - Normalization"
author: "Iago"
date: "2024-09-17"
output: html_notebook
---
# 3. Pre-normalization  
Stage description:  
  - Probes with fewer than 3 beads in >1% of samples should be removed  
  - CpG sites with missing data for > 10% of the samples should be removed  
  
```{r message=FALSE, warning=FALSE, include=FALSE, paged.print=FALSE}
source("~/walssbasslab_server/Walss-Bass lab juliet/Iago/1.EWAS_K01_Analysis/Aug.24-EWAS_BDvsHC_BRAINBANK/source_functions/source_functions.R")

# windows path
#source("V:/common/Walss-Bass_Lab/Walss-Bass lab juliet/Iago/1.EWAS_K01_Analysis/Aug.24-EWAS_BDvsHC_BRAINBANK/source_functions/source_functions.R")

need_packages <-c("tibble", "readxl","minfi", "readxl", "stringr", "dplyr", "IlluminaHumanMethylationEPICmanifest", "wateRmelon", "ewastools")#' install by running the function
install_cran_pkgs(pkgs = need_packages)
install_bioconductor_pkgs(pkgs = need_packages)

#' Check if all you need is installed
check_installed(pkgs = need_packages)

#load all packages, if needed
load_pkgs <-c("tibble", "readxl","minfi", "readxl", "stringr", "dplyr", "IlluminaHumanMethylationEPICmanifest", "wateRmelon", "ewastools")
lapply(load_pkgs, require, character.only = TRUE )
```


#### 0.1 Directory variables
```{r}
inputFolder <-  "/home/frieslab/walssbasslab_server/Walss-Bass lab juliet/Iago/1.EWAS_K01_Analysis/Aug.24-EWAS_BDvsHC_BRAINBANK"
outputFolder <- inputFolder
knitr::opts_knit$set(root.dir = inputFolder)

```

#### 0.2 Input variables 
```{r}
date <- "Aug.2024"
# - Targets
targets <- file.path(inputFolder, "6.Phenotype_files/K01_annotation_QCmetrics_removalAug.2024.csv")

# - RGset
load(file.path(inputFolder, "7.Intermediate_files/RGset_goodsamples_no-SNPs_BRAIN-BANK_Aug.2024.RData"))

# - DetP
load(file.path(inputFolder, "7.Intermediate_files/detP_BRAIN-BANK.RData"))

# - CpG missint threshold
cpg_missing_threshold <- 0.10
```
#### 0.3 Output variables 
```{r}
rgset_beads_cpgmissing_file <- file.path(inputFolder,"7.Intermediate_files/RGset_goodsamples_no-SNPs_BeadFilter_CpG10perctFilter_BRAIN-BANK_Aug.2024.RData")
```


# ------------------------------------------------------------------------------
## 3.1 Probe bead filter ----
  Stage description:
    - Checking if there are any probes that have less than 3 beads detected. 
      Type I should have two Address (A and B) and Type II should have Address A,
      comprising 3 beads.
```{r}
lb = getNBeads(RGset) < 3
pi1 = getProbeInfo(RGset, type = "I")
pi2 = getProbeInfo(RGset, type = "II")
ex1 = pi1$Name[rowMeans(lb[pi1$AddressA,] | lb[pi1$AddressB,]) > 0.01]
ex2 = pi2$Name[rowMeans(lb[pi2$AddressA,]) > 0.01]
exclude.bds = unique(c(ex1, ex2))
rm(lb, pi1, pi2, ex1, ex2)
RGset = subsetByLoci(
RGset,
excludeLoci = exclude.bds)

save(RGset, file =rgset_beads_cpgmissing_file )
```

## 3.2 Removing duplicates ----
Comparing DetP means between duplicates and removing the worst.
Stage description:
    1.) Discover which one of them has the best detP (lowest). Meaning that the sample is  
```{r}
load(file.path(inputFolder, "7.Intermediate_files/Extra-duplicates/Methylation_data_duplicates.rdata"))
targets <- read.csv(file.path(inputFolder, "6.Phenotype_files/K01_annotation_QCmetrics_removalAug.2024.csv"), row.names = 1)
dup <- colnames(RawMvalues_duplicates)

# Loading barcode and sampleID file
tmp1 <- read.csv(file.path(inputFolder, "7.Intermediate_files/Extra-duplicates/duplicates.csv"), row.names = 1)
tmp1$join <- paste(tmp1$barcode, tmp1$Sample_ID, sep = "_")

# Filtering the targets
targets_dup <- inner_join(targets[targets$Sample_ID %in% dup, ], tmp1, by = "barcode")

# Generating the RGset
load(rgset_beads_cpgmissing_file)
RGset_dup <- RGset[, colnames(RGset) %in% targets_dup$barcode]
detP_dup <- minfi::detectionP(RGset_dup)

# Sorting to get the best p-values first
detP_dup_order <- detP_dup[, order(colMeans(detP_dup))]

# Calculate column means and merge with targets_dup
col_means_detP_dup <- colMeans(detP_dup)
targets_dup <- merge(targets_dup, data.frame(barcode = names(col_means_detP_dup), col_means = col_means_detP_dup), by = "barcode")

# Order targets_dup by col_means
targets_dup <- targets_dup[order(targets_dup$col_means), ]

# Identify and save duplicates to remove
targets_remove_barcode <- targets_dup$barcode[duplicated(targets_dup$Sample_ID.x)]
save(targets_remove_barcode, file = file.path(inputFolder, "7.Intermediate_files/Extra-duplicates/DUPLICATES_TO_REMOVE.RData"))

# Plot detection p-values
print(colnames(detP_dup_order))
barplot(colMeans(detP_dup_order), las = 2, cex.names = 0.8, ylab = "Mean detection p-values")
abline(h = 0.001, col = "red")

```

```{r}
load(rgset_beads_cpgmissing_file)
load( file.path(inputFolder,"7.Intermediate_files/Extra-duplicates/DUPLICATES_TO_REMOVE.RData"))

RGset <- RGset[,which(!colnames(RGset) %in% targets_remove_barcode)]
save(RGset, file = rgset_beads_cpgmissing_file)
```


## 3.3 Identifying CpGs with more than 10% of missing values ----
```{r}

# Identify CpG sites with more than 10% missing values

MSet <- preprocessRaw(RGset)
save(MSet, file = file.path(inputFolder,"7.Intermediate_files/MSet_QC_noQNorm_BRAIN-BANK.RData"))

RawBetas <- getBeta(MSet)
RawBetas <- as.data.frame(RawBetas)
RawBetas_filtered <- RawBetas %>%
  filter(rowSums(is.na(.)) < (0.1 * ncol(RawBetas)))

RawBetas <- RawBetas_filtered

save(RawBetas, file = file.path(inputFolder,paste0("7.Intermediate_files/1.Betas/RawBetas_BRAINBANK",date,".RData")))

```

```{r}
rm(list = ls())

```

# 4. Normalization ----
Stage description:  
  - Apply quantile normalization to the latest RGset generated after QC  
  - Remove probes located on Sex chromosomes. 
  
```{r message=FALSE, warning=FALSE, include=FALSE, paged.print=FALSE}
#Fedora source path
source("~/walssbasslab_server/Walss-Bass lab juliet/Iago/1.EWAS_K01_Analysis/Aug.24-EWAS_BDvsHC_BRAINBANK/source_functions/source_functions.R")

# windows path
#source("V:/common/Walss-Bass_Lab/Walss-Bass lab juliet/Iago/1.EWAS_K01_Analysis/Aug.24-EWAS_BDvsHC_BRAINBANK/source_functions/source_functions.R")

need_packages <-c("tibble", "readxl","minfi", "readxl", "stringr", "dplyr", "IlluminaHumanMethylationEPICmanifest", "wateRmelon", "ewastools")#' install by running the function
install_cran_pkgs(pkgs = need_packages)
install_bioconductor_pkgs(pkgs = need_packages)

#' Check if all you need is installed
check_installed(pkgs = need_packages)

#load all packages, if needed
load_pkgs <-c("tibble", "readxl","minfi", "readxl", "stringr", "dplyr", "IlluminaHumanMethylationEPICmanifest", "wateRmelon", "ewastools")
lapply(load_pkgs, require, character.only = TRUE )
```


#### 0.1 Directory variables
```{r}
inputFolder <-  "/home/frieslab/walssbasslab_server/Walss-Bass lab juliet/Iago/1.EWAS_K01_Analysis/Aug.24-EWAS_BDvsHC_BRAINBANK"
outputFolder <- inputFolder
knitr::opts_knit$set(root.dir = inputFolder)
```

#### 0.2 Input variables 
```{r}
date <- "Aug.2024"
input_file <- file.path(inputFolder,"6.Phenotype_files/K01_annotation_QCmetrics_removalAug.2024.csv")
rgset_file <- file.path(inputFolder,"7.Intermediate_files/RGset_goodsamples_no-SNPs_BeadFilter_CpG10perctFilter_BRAIN-BANK_Aug.2024.RData")
RawBeta_file <- file.path(inputFolder,"7.Intermediate_files/1.Betas/RawBetas_BRAINBANKAug.2024.RData")
manifest <- file.path(inputFolder,"infinium-methylationepic-v-1-0-b5-manifest-file.csv")

```
#### 0.3 Output variables 
```{r}
QNBeta <- file.path(inputFolder,"7.Intermediate_files/RGset_goodsamples_no-SNPs_BeadFilter_CpG10perctFilter_BRAIN-BANK_Aug.2024.RData")
```
```{r}
load(rgset_file)
load(RawBeta_file)
```

# ------------------------------------------------------------------------------
### 4.1 DensityPlot RawBetas ----
  Obs: as.matrix(RawBetas) -> the function needs a matrix or RGset file, 
  them I transformed it in a Matrix
```{r}
densityPlot(as.matrix(RawBetas), sampGroups = NULL, main = "Raw Betas with quality control applied", xlab = "Beta")
```

### 4.2 Applying Quantile Normalization to RGset  ----
```{r}
QN_GRset <- preprocessQuantile(RGset)

```

## 4.2) Removing sex cromossomes and SNP-located chromossomes ----
  *Iago's notes: We are taking the sex chromosomes after normalization due to minfi preprocessQuantile() built-in "problems" dealing with 
data without sex chromosomes*

i) Subsetting sex-chromossome located CpGs
```{r}


anno <- read.csv(manifest,as.is=TRUE, skip = 7)
anno <- anno[,c('Infinium_Design_Type','Color_Channel', 'CHR', 'MAPINFO', 'Name')]

cas <- anno[substr(anno$Name, 1,3)=='ch.' & !(anno$CHR %in% c('X','Y')),]
cgs <- anno[substr(anno$Name, 1,2)=='cg'& !(anno$CHR %in% c('X','Y')),]

auto <- c(cgs$Name, cas$Name)

auto <- as.matrix(auto)


```

ii) Actually removing sex-chromossome located CpGs and SNP-located CpGs
```{r}

QNBeta <- getBeta(QN_GRset)
QNBeta_auto <- QNBeta[which(rownames(QNBeta) %in% auto),]
QNBeta <- QNBeta_auto

save(QNBeta, file = file.path(inputFolder,paste0("7.Intermediate_files/1.Betas/QNBetas_BRAINBANK_",date,".RData")))

png(filename=file.path(inputFolder,"7.Intermediate_files/1.Betas/DensityPlot-QNBetas_BRAINBANK_Aug.2024.RData"))
densityPlot(as.matrix(QNBeta), sampGroups = NULL, main = "QN Betas  with quality control applied", xlab = "Beta")
dev.off()

```

iii) Generating M-Values
```{r}
RawMvalues <- RGset %>% preprocessRaw() %>% mapToGenome() %>% getM()

RawMvalues<- RawMvalues[which(rownames(RawMvalues) %in% auto),]
dim(RawMvalues)
save(RawMvalues, file = file.path(inputFolder,paste0("7.Intermediate_files/1.Betas/RawMvalues_BRAINBANK_",date,".RData")))


```

iv) Generating M-values from QN GRset 
```{r}
QNMValues <- QN_GRset %>% getM()
QNMValues<- QNMValues[which(rownames(QNMValues) %in% auto),]
dim(QNMValues)
save(QNMValues, file = file.path(inputFolder,paste0("7.Intermediate_files/1.Betas/QNMvalues_BRAINBANK_",date,".RData")))
```

Obs: If needed, this piece of code can count how many probes do sex chromossomes
have.
```{r}
### Obs: Counting how many probes we have in sex chromossmes

# casXY <- anno[!substr(anno$Name, 1,3)=='ch.' & (anno$CHR %in% c('X','Y')),]
# cgsXY <- anno[!substr(anno$Name, 1,2)=='cg'& (anno$CHR %in% c('X','Y')),]
# 
# XYchr <- rbind(cgsXY,casXY)
```

### Saving RGSet ----
  Saving RGset with only autossome-located probes included.

```{r}

save(RGset, file = file.path(inputFolder,"7.Intermediate_files/RGset_goodsamples_autossomes-noSNP_BRAIN-BANK_Aug.2024.RData"))
```

```{r}
cat("end of script")
```